ledger/blockqueue.go

```diff
@@ -56,51 +56,68 @@ func bqInit(l *Ledger) (*blockQueue, error) {
 	bq := &blockQueue{}
 	bq.cond = sync.NewCond(&bq.mu)
 	bq.l = l
-	bq.running = true
-	bq.closed = make(chan struct{})
-	ledgerBlockqInitCount.Inc(nil)
-	start := time.Now()
-	err := bq.l.blockDBs.Rdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
-		var err0 error
-		bq.lastCommitted, err0 = blockdb.BlockLatest(tx)
-		return err0
-	})
-	ledgerBlockqInitMicros.AddMicrosecondsSince(start, nil)
-	if err != nil {
-		return nil, err
-	}
-
-	go bq.syncer()
 	return bq, nil
 }
 
-func (bq *blockQueue) close() {
+func (bq *blockQueue) start() error {
 	bq.mu.Lock()
-	defer func() {
-		bq.mu.Unlock()
-		// we want to block here until the sync go routine is done.
-		// it's not (just) for the sake of a complete cleanup, but rather
-		// to ensure that the sync goroutine isn't busy in a notifyCommit
-		// call which might be blocked inside one of the trackers.
-		<-bq.closed
-	}()
+	defer bq.mu.Unlock()
+
+	if !bq.running {
+		if bq.closed != nil {
+			// a previus close() is still waiting on a previous syncer() to finish
+			oldsyncer := bq.closed
+			bq.mu.Unlock()
+			<-oldsyncer
+			bq.mu.Lock()
+		}
+		bq.running = true
+		bq.closed = make(chan struct{})
+		ledgerBlockqInitCount.Inc(nil)
+		start := time.Now()
+		err := bq.l.blockDBs.Rdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+			var err0 error
+			bq.lastCommitted, err0 = blockdb.BlockLatest(tx)
+			return err0
+		})
+		ledgerBlockqInitMicros.AddMicrosecondsSince(start, nil)
+		if err != nil {
+			return err
+		}
 
+		go bq.syncer()
+	}
+	return nil
+}
+
+func (bq *blockQueue) close() {
+	bq.mu.Lock()
+	closechan := bq.closed
 	if bq.running {
 		bq.running = false
 		bq.cond.Broadcast()
 	}
-
+	bq.mu.Unlock()
+
+	// we want to block here until the sync go routine is done.
+	// it's not (just) for the sake of a complete cleanup, but rather
+	// to ensure that the sync goroutine isn't busy in a notifyCommit
+	// call which might be blocked inside one of the trackers.
+	if closechan != nil {
+		<-closechan
+	}
 }
 
 func (bq *blockQueue) syncer() {
-	defer close(bq.closed)
```

Comment:
The new behavior will not close `dq.closed` when `syncer` returns because of an error. Why?